### ISSUE TITLE
Issues with missing shared object dependencies when running in aws environment

### ISSUE THREAD

**Issue by jhrcek** [seek-oss/serverless-haskell#149]
Hello,

We're hitting some issues with dynamic library dependencies of a lambda deployed via this library.
I included as much details as I could gather (sorry if it's too much). Any help on this would be appreciated :smiley: 

When I deploy the lambda by running `npx sls deploy` and do a test run of it in AWS, I'm getting the following error:

```
START RequestId: 07ef6bc4-ef85-49bf-82eb-8a73a1e313c3 Version: $LATEST
./r-executor-cleanup: /lib64/libldap_r-2.4.so.2: no version information available (required by /var/task/libpq.so.5)
./r-executor-cleanup: /lib64/libm.so.6: version `GLIBC_2.23' not found (required by /var/task/libquadmath.so.0)
END RequestId: 07ef6bc4-ef85-49bf-82eb-8a73a1e313c3
REPORT RequestId: 07ef6bc4-ef85-49bf-82eb-8a73a1e313c3 Duration: 26.71 ms Billed Duration: 100 ms Memory Size: 1024 MB Max Memory Used: 6 MB
RequestId: 07ef6bc4-ef85-49bf-82eb-8a73a1e313c3 Error: Runtime exited with error: exit status 1 Runtime.ExitError
```

The lambda looks like this so far (but the binary has all our required haskell dependencies)
```haskell
module Our.Module.Name where
import AWSLambda (lambdaMain)
import Data.Aeson (Value)

lambda :: IO ()
lambda = lambdaMain cleanupHandler

cleanupHandler :: Value -> IO ()
cleanupHandler _ = putTextLn "TODO implement me"
```

## build environment
We tested building the lambda in couple of environments, each time the error we got when running Test in aws was the same:

- Linux 5.8.8, Fedora 32
- Linux 5.4.62, NixOS 21.03
- MacOS

We're building our project using stack lts-16.2,
we tried both with serverless-haskell 0.11.3 (included in the [lts-16.2](https://www.stackage.org/lts-16.2/package/serverless-haskell-0.11.3) snapshot) as well as with the latest version from hackage (0.12.1 - added to stack.yaml inside `extra-deps`)
Our package.json looks like this:
<details>
  <summary>package.json</summary>
  <pre>
{
  "name": "pi-haskell-lambdas",
  "version": "0.1.0",
  "description": "Config for building and deploying haskell lambdas via serverless framework",
  "directories": {
    "test": "test"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "serverless": "^1.83.0",
    "serverless-haskell": "^0.12.1"
  }
}
  </pre>
</details>

## Addition info + things we tried:

As for the two .so files mentioned in the error log:
libpq.so.5 - is dependency of postgresql-simple that we're using
libquadmath.so.0  - this is dependency of some of the linear algebra / machine learning libs our app is using

<details>
  <summary>Which files end up in the .serverless/r-executor-cleanup.zip ?</summary>
  <pre>
$ unzip .serverless/r-executor-cleanup.zip
$ ldd r-executor-cleanup  # This is our  main binary packaged in the zip
	linux-vdso.so.1 (0x00007ffc1b37f000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fedba2e3000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fedba2c1000)
	libpq.so.5 => /lib64/libpq.so.5 (0x00007fedba270000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007fedba26b000)
	libblas.so.3 => /lib64/libblas.so.3 (0x00007fedba214000)
	liblapack.so.3 => /lib64/liblapack.so.3 (0x00007fedb9b69000)
	libgsl.so.19 => not found
	libgslcblas.so.0 => /lib64/libgslcblas.so.0 (0x00007fedb9b24000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fedb9b0a000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fedb9aff000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fedb9af8000)
	libgmp.so.10 => /lib64/libgmp.so.10 (0x00007fedb9a61000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fedb9895000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fedba449000)
	libssl.so.1.1 => /lib64/libssl.so.1.1 (0x00007fedb97fe000)
	libcrypto.so.1.1 => /lib64/libcrypto.so.1.1 (0x00007fedb9511000)
	libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007fedb94ba000)
	libldap_r-2.4.so.2 => /lib64/libldap_r-2.4.so.2 (0x00007fedb9460000)
	libgfortran.so.5 => /lib64/libgfortran.so.5 (0x00007fedb9195000)
	libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007fedb90a8000)
	libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007fedb908f000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007fedb9088000)
	libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007fedb9076000)
	libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007fedb906f000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007fedb9055000)
	liblber-2.4.so.2 => /lib64/liblber-2.4.so.2 (0x00007fedb9042000)
	libsasl2.so.3 => /lib64/libsasl2.so.3 (0x00007fedb9022000)
	libquadmath.so.0 => /lib64/libquadmath.so.0 (0x00007fedb8fd8000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fedb8fbd000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fedb8f90000)
	libcrypt.so.2 => /lib64/libcrypt.so.2 (0x00007fedb8f53000)
	libpcre2-8.so.0 => /lib64/libpcre2-8.so.0 (0x00007fedb8eba000)
  </pre>
</details>

<details>
  <summary>What version of GLIBC do the ".so" files depend on?</summary>
  <pre>
$ for binary in r-executor-cleanup *.so.*; do echo $binary; objdump -T $binary | grep -o -E 'GLIBC_[0-9.]+' | sort -u; done
r-executor-cleanup
GLIBC_2.10
GLIBC_2.12
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.2
GLIBC_2.3.4
GLIBC_2.7
GLIBC_2.8
libasn1.so.8
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.8
libblas.so.3
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
libcrypto.so.1.0.0
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.7
libcrypt.so.1
GLIBC_2.14
GLIBC_2.2.5
libgfortran.so.3
GLIBC_2.14
GLIBC_2.17
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.6
GLIBC_2.7
libgnutls.so.30
GLIBC_2.14
GLIBC_2.15
GLIBC_2.17
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.2
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.8
libgslcblas.so.0
GLIBC_2.2.5
GLIBC_2.3.4
libgsl.so.19
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.7
libgssapi.so.3
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.4
GLIBC_2.8
libhcrypto.so.4
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.7
libheimbase.so.1
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.8
libheimntlm.so.0
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.4
libhogweed.so.4
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.4
libhx509.so.5
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.8
libidn.so.11
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.4
libkrb5.so.26
GLIBC_2.14
GLIBC_2.15
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.2
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.7
GLIBC_2.8
liblapack.so.3
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.4
liblber-2.4.so.2
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
libldap_r-2.4.so.2
GLIBC_2.12
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.2
GLIBC_2.3.4
GLIBC_2.4
libnettle.so.6
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3.4
GLIBC_2.4
libpq.so.5
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
libquadmath.so.0
GLIBC_2.10
GLIBC_2.14
GLIBC_2.23
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
libroken.so.18
GLIBC_2.14
GLIBC_2.15
GLIBC_2.16
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
GLIBC_2.8
libsasl2.so.2
GLIBC_2.14
GLIBC_2.15
GLIBC_2.2.5
GLIBC_2.3
GLIBC_2.3.4
GLIBC_2.4
libsqlite3.so.0
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3.4
GLIBC_2.4
libssl.so.1.0.0
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3.4
GLIBC_2.4
libwind.so.0
GLIBC_2.14
GLIBC_2.2.5
GLIBC_2.3.4
GLIBC_2.4

  </pre>
</details>

### Where do the .so files, that end up in the zip package come from?
I added some logging [between these 2 lines](https://github.com/seek-oss/serverless-haskell/blob/50fc0a065ab942f03eee1b094dce10802ccdddc5/index.ts#L354-L355) and this is what I get as part of `npx sls package`:

<details>
  <summary>output</summary>
  <pre>
$ npx sls package
Serverless: Building handler r-executor-cleanup with Stack...
# ...
# stack build log omitted
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libpq.so.5 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libpq.so.5
Serverless: Running stack exec cp /usr/lib/libblas.so.3 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libblas.so.3
Serverless: Running stack exec cp /usr/lib/liblapack.so.3 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/liblapack.so.3
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libgsl.so.19 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libgsl.so.19
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libgslcblas.so.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libgslcblas.so.0
Serverless: Running stack exec cp /lib/x86_64-linux-gnu/libssl.so.1.0.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libssl.so.1.0.0
Serverless: Running stack exec cp /lib/x86_64-linux-gnu/libcrypto.so.1.0.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libcrypto.so.1.0.0
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libldap_r-2.4.so.2 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libldap_r-2.4.so.2
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libgfortran.so.3 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libgfortran.so.3
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/liblber-2.4.so.2 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/liblber-2.4.so.2
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libsasl2.so.2 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libsasl2.so.2
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libgssapi.so.3 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libgssapi.so.3
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libgnutls.so.30 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libgnutls.so.30
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libquadmath.so.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libquadmath.so.0
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libheimntlm.so.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libheimntlm.so.0
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libkrb5.so.26 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libkrb5.so.26
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libasn1.so.8 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libasn1.so.8
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libhcrypto.so.4 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libhcrypto.so.4
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libroken.so.18 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libroken.so.18
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libidn.so.11 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libidn.so.11
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libnettle.so.6 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libnettle.so.6
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libhogweed.so.4 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libhogweed.so.4
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libwind.so.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libwind.so.0
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libheimbase.so.1 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libheimbase.so.1
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libhx509.so.5 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libhx509.so.5
Serverless: Running stack exec cp /usr/lib/x86_64-linux-gnu/libsqlite3.so.0 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libsqlite3.so.0
Serverless: Running stack exec cp /lib/x86_64-linux-gnu/libcrypt.so.1 /home/jhrcek/Devel/github.com/Holmusk/PI/backend/libcrypt.so.1
Serverless: Packaging service...
  </pre>
</details>

## The issue is reproducible locally

When I run `npx sls invoke local -f r-executor-cleanup` locally
I see an active docker container, which seems to be doing nothing (for 30 minutes).

```bash
docker ps -a --no-trunc
CONTAINER ID                                                       IMAGE                 COMMAND                                         CREATED             STATUS              PORTS               NAMES
cfdb5dcd255da7edbbdd25cc0f5769aa6f99278493406df80befc0003e7d93a4   sls-docker-provided   "/var/runtime/init ./r-executor-cleanup \"\""   58 seconds ago      Up 57 seconds                           distracted_blackburn
```

And when I do `docker kill <CONTAINER_ID>` the following appears in terminal, where I executed the above command:

<details>
  <summary>output</summary>
  <pre>
Serverless: Packaging service...
# At this point I execute `docker kill...`
./r-executor-cleanup: /lib64/libldap_r-2.4.so.2: no version information available (required by /var/task/libpq.so.5)
./r-executor-cleanup: /lib64/libm.so.6: version `GLIBC_2.23' not found (required by /var/task/libquadmath.so.0)
 
  Error --------------------------------------------------
 
  Error: Failed to run docker for provided image (exit code 137})
      at /home/jhrcek/Devel/github.com/Holmusk/PI/backend/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:537:21
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Get Support --------------------------------------------
     Docs:          docs.serverless.com
     Bugs:          github.com/serverless/serverless/issues
     Issues:        forum.serverless.com
 
  Your Environment Information ---------------------------
     Operating System:          linux
     Node Version:              12.16.1
     Framework Version:         1.83.0 (local)
     Plugin Version:            3.8.4
     SDK Version:               2.3.1
     Components Version:        2.34.9
  </pre>
</details>

Note that this corresponds to the error we're getting when testing the lambda in AWS console.

**Comment by koterpillar** [comment]
Hi, and thanks for the very detailed writeup. Could you please show your `serverless.yml` for completeness?

**Comment by jhrcek** [comment]
S
<details>
  <summary>serverless.yaml</summary>
  <pre>
service: lambda

provider:
  name: aws
  region: ap-southeast-1
  runtime: haskell

functions:
  r-executor-cleanup:
    handler: pi.r-executor-cleanup

package:
  individually: true
  exclude:
    - artifacts/**
    - data/**
    - dist/**
    - dist-newstyle/**
    - node_modules/**
    - .serverless/**
    - .stack-work/**

plugins:
  - serverless-haskell

  </pre>
</details>

**Comment by koterpillar** [comment]
I believe your issue will be fixed by #150. Can you try that branch (or change the runtime constant to `provided.al2` locally)?

**Comment by jhrcek** [comment]
Thank you very much for looking into this :smile: 
Applied that change locally and it's working now both in AWS console as well as with `npx sls invoke local`.

**Comment by koterpillar** [comment]
Released as 0.12.2.
